packages/jest-haste-map/src/crawlers/node.ts

```diff
@@ -32,22 +32,42 @@ function find(
 
   function search(directory: string): void {
     activeCalls++;
-    fs.readdir(directory, (err, names) => {
+    fs.readdir(directory, {withFileTypes: true}, (err, entries) => {
       activeCalls--;
       if (err) {
         callback(result);
         return;
       }
-      names.forEach(file => {
-        file = path.join(directory, file);
-        if (ignore(file)) {
-          return;
+      entries.forEach((entry: string | fs.Dirent) => {
+        let file: string;
+        // node < v10.10 does not support the withFileTypes option, and
+        // entry will be a string.
+        if (typeof entry === 'string') {
+          file = path.join(directory, entry);
+        } else {
+          file = path.join(directory, entry.name);
+
+          if (ignore(file)) {
```

Comment:
instead of these `if`s, could we extract the logic from within the `lstat` calls into a private method on the class and reuse the logic?

Reply:
@SimenB Hmm, it's a little tricky to do this. There are three pieces of logic, and depending on the node version they can be done with the readdir result (v10.10+) or after lstat:
1. Ignoring files by name - always possible with readdir result
2. abort/recurse if file is not a normal file - readdir in v10.10+, lstat otherwise
3. Append file to results array - this requires the file mtime, which is only available via lstat.

Supposing there is a shared function:

1. #1 should always be outside of it, since we always want to do it before lstat
2. #3 should also be excluded because it always has to be inside lstat.

So really we're just sharing the logic for #2, but that's a little confusing to put in a function because it both returns a boolean (should we continue with attempting to process as a normal file) as well as performs an imperative action (recurse if file is a directory).

Reply:
ok, thanks for the detail. What about just setting a `inspectedStat` or whatever based on `typeof entry === 'string'` which skips the part you've commented `// This logic is unnecessary for node > v10.10` etc on? I think it'd just be easier to follow the logic than anything else

Reply:
@SimenB Got it, I took a stab at it in the PR - I'm not sure if I like it (it's actually *more* code), but maybe you have some better ideas. Happy to make any changes to get this pr merged!

Reply:
yeah agreed, that didn't turn out any cleaner after all. Let's just revert that last commit 👍